### PR TITLE
Verify resolution of Ivy/Maven published files

### PR DESCRIPTION
**Pull request by melix**
### Context

See #3671

### CODE REVIEW

**Comment by melix** [comment]
@bigdaz I wanted to remove this:

```
        else {
            executer.beforeExecute {
                // Remove the experimental flag set earlier...
                // TODO:DAZ Remove this once we support excludes and we can have a single flag to enable publish/resolve
                withArguments()
            }
        }
```

from `doResolveArtifacts` as I think it's redundant now (because I use `gradleMetadata()` or `mavenPom()` in metadata sources).

However if I do something strange happens:

`IvyPublishJavaIntegTest#"can publish java-library with rejected versions"` starts to fail. It fails with:

```
assert resolutionResult == expectedFileNames
       |                |  |
       |                |  [commons-collections-3.2.2.jar, commons-logging-1.1.1.jar, publishTest-1.9.jar, spring-core-2.5.6.jar]
       |                false
       [commons-collections-3.2.jar, commons-logging-1.1.1.jar, publishTest-1.9.jar, spring-core-2.5.6.jar]
```

which means that it takes the reject clauses into consideration when resolving without Gradle metadata. We can also see that the `resolveArtifact` task is `UP-TO-DATE`.

I have checked that:

- the signature for the repository is different (it is, thanks to the sources `appendId`) for the 2 resolution rounds
- the Gradle metadata source is **not** used for the 2d round (that checks Ivy)
- that if I force `outputs.upToDateWhen { false }`, I still have the error

So I'm out of ideas. Note that the Maven test fails in a similar way. Those are the only 2 tests failing, so maybe I'm missing something obvious.

**Comment by bigdaz** [comment]
>which means that it takes the reject clauses into consideration when resolving without Gradle metadata

Won't this be due to the fact that we don't use the configured `metadataSources` for version listing?

File: subprojects/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishJavaIntegTest.groovy [hunk 1]
```diff
@@ -240,8 +243,21 @@ class IvyPublishJavaIntegTest extends AbstractIvyPublishIntegTest {
...
 
         and:
-        resolveArtifacts(javaLibrary) == ["commons-collections-3.2.2.jar", "commons-io-1.4.jar", "publishTest-1.9.jar"]
-        resolveAdditionalArtifacts(javaLibrary) == ["publishTest-1.9-source.jar"]
+        resolveArtifacts(javaLibrary) {
+            withoutModuleMetadata {
+                expectFiles "commons-collections-3.2.2.jar", "commons-io-1.4.jar", "publishTest-1.9.jar", "publishTest-1.9-source.jar"
+            }
+            withModuleMetadata {
+                // TODO: handle additional files in Gradle metadata?
```
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishJavaIntegTest.groovy]: Replaced with a simple comment

File: subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishArtifactCustomizationIntegTest.groovy [hunk 2]
```diff
@@ -75,7 +100,16 @@ class MavenPublishArtifactCustomizationIntegTest extends AbstractMavenPublishInt
         module.assertArtifactsPublished("projectText-1.0-classified.jar", "projectText-1.0.pom")
 
         and:
-        resolveArtifact(module, 'jar', 'classified') == ["projectText-1.0-classified.jar"]
+        resolveArtifacts(module) {
+            classifier = 'classified'
+            ext = 'jar'
+            withModuleMetadata {
+                shouldFail()
```
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishArtifactCustomizationIntegTest.groovy]: Issue https://github.com/gradle/gradle/issues/3761

File: subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishArtifactCustomizationIntegTest.groovy [hunk 3]
```diff
@@ -105,7 +139,15 @@ class MavenPublishArtifactCustomizationIntegTest extends AbstractMavenPublishInt
         module.assertArtifactsPublished("projectText-1.0.pom", "projectText-1.0.txt", "projectText-1.0-customjar.jar")
 
         and:
-        resolveArtifacts(module, [classifier: 'customjar']) == ["projectText-1.0-customjar.jar", "projectText-1.0.txt"]
+        resolveArtifacts(module) {
+            classifier = 'customjar'
+            withModuleMetadata {
+                shouldFail()
```
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishArtifactCustomizationIntegTest.groovy]: I have removed the simple `shouldFail()` version. Now we are forced to use the `shouldFail { ... }` which I hope is an incentive to explain why.

**Pull request merged**